transport/grpc/config_test.go

```diff
@@ -353,12 +375,67 @@ func TestTransportSpec(t *testing.T) {
 					dialer, ok := single.Transport().(*Dialer)
 					require.True(t, ok, "expected *Dialer, got %T", single.Transport())
 					assert.Equal(t, wantOutbound.TLS, dialer.options.creds != nil)
+					if wantOutbound.WantCustomContextDialer {
+						assert.NotNil(t, dialer.options.contextDialer, "expected custom context dialer")
+					}
 				}
 			}
 		})
 	}
 }
 
+func TestContextDialerOptionUsage(t *testing.T) {
+	type attrs map[string]interface{}
+
+	lis, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	defer lis.Close()
+	server := grpc.NewServer()
+	defer server.Stop()
+	go func() {
+		require.NoError(t, server.Serve(lis))
+	}()
+
+	dialContextInvoked := int32(0)
+	dialer := func(ctx context.Context, addr string) (net.Conn, error) {
+		atomic.AddInt32(&dialContextInvoked, 1)
+		return (&net.Dialer{}).DialContext(ctx, "tcp", addr)
+	}
+	configurator := yarpcconfig.New()
+	err = configurator.RegisterTransport(TransportSpec(ContextDialer(dialer)))
+	require.NoError(t, err)
+	cfgData := attrs{
+		"outbounds": attrs{
+			"myservice": attrs{
+				TransportName: attrs{"address": lis.Addr().String()},
+			},
+		},
+	}
+	cfg, err := configurator.LoadConfig("foo", cfgData)
+	require.NoError(t, err)
+	outbound, ok := cfg.Outbounds["myservice"].Unary.(*Outbound)
+	require.True(t, ok)
+	defer outbound.Stop()
+	single := outbound.peerChooser.(*peer.Single)
+	require.NoError(t, single.Start())
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	peer, _, err := single.Choose(ctx, &transport.Request{})
+	require.NoError(t, err)
+	grpcPeer, ok := peer.(*grpcPeer)
+	require.True(t, ok)
+
+	for {
+		state := grpcPeer.clientConn.GetState()
+		if grpcPeer.clientConn.GetState() == connectivity.Ready {
+			break
+		}
+		grpcPeer.clientConn.WaitForStateChange(ctx, state)
+	}
+	require.Equal(t, connectivity.Ready, grpcPeer.clientConn.GetState())
+	require.Equal(t, int32(1), dialContextInvoked)
```

Comment:
add a bit context on why we are asserting 1: `require.Equal(t, int32(1), dialContextInvoked, "counter should increment by one from option invocation")`